Change getTrackInfo to use getNode

low-batt opened the pull request:
This commit will:
- Change the `getTrackInfo` method in the `PlayerCore` class to get all tracks at once using `getNode`
- Add a `nodeValueAsInt` method to `MPVController` for extracting an integer from a node map

Instruments identified `getTrackInfo` as sometimes being responsible for short hangs. This is because `mpv_get_property` sometimes has to wait to acquire a mpv lock. The current `getTrackInfo` implementation makes 20 calls to `mpv_get_property` for each track. By changing `getTrackInfo` to make one call to `getNode`, obtaining the entire list in a single call to `mpv_get_property`, the time waiting for `mpv_get_property` is reduced.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [ ] This implements/fixes issue #.

---

**Description:**

low-batt committed:
Change getTrackInfo to use getNode

This commit will:
- Change the getTrackInfo method in the PlayerCore class to get all
  tracks at once using getNode
- Add a nodeValueAsInt method to MPVController for extracting an integer
  from a node map

Instruments identified getTrackInfo as sometimes being responsible for
short hangs. This is because mpv_get_property sometimes has to wait to
acquire a mpv lock. The current getTrackInfo implementation makes 20
calls to mpv_get_property for each track. By changing getTrackInfo to
make one call to getNode, obtaining the entire list in a single call to
mpv_get_property, the time waiting for mpv_get_property is reduced.

uiryuu commented on iina/PlayerCore.swift:
```diff
@@ -2816,30 +2681,38 @@ class PlayerCore: NSObject {
+          continue
+        }
+        let id = MPVController.nodeValueAsInt(dict["id"])
+        let track = MPVTrack(id: id, type: MPVTrack.TrackType(rawValue: type)!,
```
It would be cleaner to write a new constructor for `MPVTrack` that accept a dict[Stirng: Any]. That way we can move all `MPVTrack` initialization logic to its class.

low-batt replied:
Agreed. Will work on that.

From what I recall I decided not to do that because I wanted to minimize the code changes as I was hoping this change would go into 1.4.2. But as this is critical code and the change was not required to fix a defect, lhc wanted to postpone this one until 1.5.0 where it can get more testing.

uiryuu reviewed (changes requested):
Rebased and tested, no problem found. See comments below. Other changes look good to me.

low-batt committed:
Change getTrackInfo to use getNode

This commit will:
- Change the getTrackInfo method in the PlayerCore class to get all
  tracks at once using getNode
- Add an initializer to MPVTrack that accepts a dictionary of properties
- Add a nodeValueAsInt method to MPVController for extracting an integer
  from a node map

Instruments identified getTrackInfo as sometimes being responsible for
short hangs. This is because mpv_get_property sometimes has to wait to
acquire a mpv lock. The current getTrackInfo implementation makes 20
calls to mpv_get_property for each track. By changing getTrackInfo to
make one call to getNode, obtaining the entire list in a single call to
mpv_get_property, the time waiting for mpv_get_property is reduced.

low-batt commented:
Updated.

In general, In this area of IINA (including `MPVController`) the code is loose, rather than tight/strict. Calls into mpv do not check the return code for an error. Properties that are not of the expected type quietly result in `nil` values. I did not try and tighten up the code and followed the existing conventions.
